<h1>Feature Documentation Notebook</h1>
<h4 style=color:blue>This is an extensive documentation of cleaning decisions and imputation decisions of all the features</h4>

In [1]:
import pandas as pd
import numpy as np
from autoviz.AutoViz_Class import AutoViz_Class
import matplotlib.pyplot as plt
import seaborn as sns
from IPython.display import display 
pd.options.display.max_columns = None

Imported v0.1.905. Please call AutoViz in this sequence:
    AV = AutoViz_Class()
    %matplotlib inline
    dfte = AV.AutoViz(filename, sep=',', depVar='', dfte=None, header=0, verbose=1, lowess=False,
               chart_format='svg',max_rows_analyzed=150000,max_cols_analyzed=30, save_plot_dir=None)


In [61]:
df = pd.read_csv("Lead Scoring.csv")

In [60]:
# We are doing AutoEda with the help of AutoViz

AV = AutoViz_Class()


av_result = AV.AutoViz(
    filename="Lead Scoring.csv",
    sep=",",
    depVar="Converted",       # Leave blank if you're just exploring, or specify target column name
    dfte=None,       # Not passing DataFrame directly
    header=0,
    verbose=1,
    lowess=False,
    chart_format='svg',
    max_rows_analyzed=150000,
    max_cols_analyzed=50
)


Shape of your Data Set loaded: (9240, 37)
#######################################################################################
######################## C L A S S I F Y I N G  V A R I A B L E S  ####################
#######################################################################################
Classifying variables in data set...
    Number of Numeric Columns =  4
    Number of Integer-Categorical Columns =  1
    Number of String-Categorical Columns =  15
    Number of Factor-Categorical Columns =  0
    Number of String-Boolean Columns =  9
    Number of Numeric-Boolean Columns =  0
    Number of Discrete String Columns =  0
    Number of NLP String Columns =  0
    Number of Date Time Columns =  0
    Number of ID Columns =  2
    Number of Columns to Delete =  5
    36 Predictors classified...
        7 variable(s) removed since they were ID or low-information variables
        List of variables removed: ['Prospect ID', 'Lead Number', 'Magazine', 'Receive More Updates Ab

Total Number of Scatter Plots = 10
All Plots done
Time to run AutoViz = 16 seconds 

 ###################### AUTO VISUALIZATION Completed ########################


In [6]:
df.head()

Prospect ID               Lead Number       Lead Origin         \
0  7927b2df-8bba-4d29-b9a2-b6e0beafe620    660737                         API   
1  2a272436-5132-4136-86fa-dcc88c88f482    660728                         API   
2  8cc8c611-a219-4f35-ad23-fdfd2656bd8a    660727     Landing Page Submission   
3  0cc2df48-7cf4-4e39-9de9-19797f9b38cc    660719     Landing Page Submission   
4  3256f628-e534-4826-9d63-4a8b88782852    660681     Landing Page Submission   

    Lead Source   Do Not Email Do Not Call  Converted  TotalVisits  \
0      Olark Chat      No           No          0          0.0       
1  Organic Search      No           No          0          5.0       
2  Direct Traffic      No           No          1          2.0       
3  Direct Traffic      No           No          0          1.0       
4          Google      No           No          1          2.0       

   Total Time Spent on Website  Page Views Per Visit      Last Activity        \
0                0                       0.0          Page Visited on Website   
1              674                       2.5                     Email Opened   
2             1532                       2.0                     Email Opened   
3              305                       1.0                      Unreachable   
4             1428                       1.0                Converted to Lead   

  Country      Specialization      How did you hear about X Education  \
0     NaN                   Select                   Select             
1   India                   Select                   Select             
2   India  Business Administration                   Select             
3   India    Media and Advertising            Word Of Mouth             
4   India                   Select                    Other             

  What is your current occupation  \
0            Unemployed             
1            Unemployed             
2               Student             
3            Unemployed             
4            Unemployed             

  What matters most to you in choosing a course Search Magazine  \
0             Better Career Prospects             No      No      
1             Better Career Prospects             No      No      
2             Better Career Prospects             No      No      
3             Better Career Prospects             No      No      
4             Better Career Prospects             No      No      

  Newspaper Article X Education Forums Newspaper Digital Advertisement  \
0         No                No             No              No            
1         No                No             No              No            
2         No                No             No              No            
3         No                No             No              No            
4         No                No             No              No            

  Through Recommendations Receive More Updates About Our Courses  \
0            No                             No                     
1            No                             No                     
2            No                             No                     
3            No                             No                     
4            No                             No                     

                  Tags                    Lead Quality    \
0          Interested in other courses  Low in Relevance   
1                              Ringing               NaN   
2  Will revert after reading the email          Might be   
3                              Ringing          Not Sure   
4  Will revert after reading the email          Might be   

  Update me on Supply Chain Content Get updates on DM Content   Lead Profile   \
0                 No                            No                     Select   
1                 No                            No                     Select   
2                 No                            No             Potential Lead   
3               

<h3>Information about all the rows and their data types </h3>

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9240 entries, 0 to 9239
Data columns (total 37 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   Prospect ID                                    9240 non-null   object 
 1   Lead Number                                    9240 non-null   int64  
 2   Lead Origin                                    9240 non-null   object 
 3   Lead Source                                    9204 non-null   object 
 4   Do Not Email                                   9240 non-null   object 
 5   Do Not Call                                    9240 non-null   object 
 6   Converted                                      9240 non-null   int64  
 7   TotalVisits                                    9103 non-null   float64
 8   Total Time Spent on Website                    9240 non-null   int64  
 9   Page Views Per Visit                           9103 

<h4>Column 0 : Prospect ID</h4>

In [8]:
len(df["Prospect ID"].unique())

9240

We get to know Prospect ID is a uniquely generated Id , Count is 9240 which is same as the total rows. Action : drop

<h4>Column 1 : Lead Number</h4>

In [9]:
len(df["Lead Number"].unique())

9240

We get to know Lead Number is a uniquely generated Id , Count is 9240 which is same as the total rows. Action : drop

<h4>Column 2 : Lead Origin</h4>

In [10]:
df["Lead Origin"].value_counts()

Lead Origin
Landing Page Submission    4886
API                        3580
Lead Add Form               718
Lead Import                  55
Quick Add Form                1
Name: count, dtype: int64

Rare columns have been detected , Make sure they are handled or try to take a decision based on it. Missing values 0

<p>Action : Try to handle the rare values , Useful column</p>


<h4>Column 3 : Lead Source</h4>

In [11]:
df["Lead Source"].isna().sum()

np.int64(36)

In [12]:
df["Lead Source"].value_counts()

Lead Source
Google               2868
Direct Traffic       2543
Olark Chat           1755
Organic Search       1154
Reference             534
Welingak Website      142
Referral Sites        125
Facebook               55
bing                    6
google                  5
Click2call              4
Live Chat               2
Social Media            2
Press_Release           2
Pay per Click Ads       1
blog                    1
WeLearn                 1
welearnblog_Home        1
youtubechannel          1
testone                 1
NC_EDM                  1
Name: count, dtype: int64

Multiple rare columns detected , Column sounds Useful. Missing values 36

Actions : Replace Na values with Others and Club Sources that can be clubbed . For example (Google,bing as Search Engine)

<h4>Column 4 : Do Not Email </h4>

In [13]:
df['Do Not Email'].isna().sum()

np.int64(0)

In [14]:
df['Do Not Email'].value_counts()

Do Not Email
No     8506
Yes     734
Name: count, dtype: int64

Heavily skewed Towards No, Apart from that no issues. 0 Missing Values

Actions : One Hot encoding

<h4>Column 5 : Do Not Call </h4>

In [15]:
df['Do Not Call'].isna().sum()

np.int64(0)

In [16]:
df['Do Not Call'].value_counts()

Do Not Call
No     9238
Yes       2
Name: count, dtype: int64

Very Heavily Skewed. Only 2 Yes out of all rows. 

Actions :  Drop ,  if in future we get more re engineer pipeline.

<h4>Column 6 : Converted <h4>

In [17]:
df['Converted'].isna().sum()

np.int64(0)

In [18]:
df['Converted'].value_counts()

Converted
0    5679
1    3561
Name: count, dtype: int64

Target Column , Light skew towards 0. 

Action : Nothing

<h4>Column 7 : TotalVisits</h4>

In [19]:
df["TotalVisits"].isna().sum()

np.int64(137)

In [20]:
((df["TotalVisits"].isna().sum())/len(df))*100

np.float64(1.4826839826839828)

In [21]:
len(df["TotalVisits"].unique())

42

In [22]:
df[df["Total Time Spent on Website"]==0]["TotalVisits"].value_counts()

TotalVisits
0.0    2189
2.0       2
5.0       1
6.0       1
Name: count, dtype: int64

In [23]:
df[df["TotalVisits"].isna()].head()

Prospect ID               Lead Number  Lead Origin    \
77   895d4905-f534-4f18-915b-8d239a72b5dc    659722     Lead Add Form   
79   3a0ce10f-d2c1-4213-a2bc-4f97bcd29699    659710     Lead Add Form   
81   277ad6a6-4565-4a18-a1ff-e46e03f22663    659705     Lead Add Form   
88   68f496c2-0073-470f-9c3c-7fb48f060ce5    659631     Lead Add Form   
120  144807db-2895-4002-b52e-3eda79c22395    659283     Lead Add Form   

       Lead Source    Do Not Email Do Not Call  Converted  TotalVisits  \
77   Welingak Website      No           No          1         NaN        
79          Reference      No           No          1         NaN        
81   Welingak Website      No           No          1         NaN        
88          Reference      No           No          1         NaN        
120         Reference      No           No          1         NaN        

     Total Time Spent on Website  Page Views Per Visit Last Activity Country  \
77              1096                      NaN                NaN        NaN    
79               705                      NaN                NaN        NaN    
81               346                      NaN                NaN        NaN    
88              1007                      NaN                NaN        NaN    
120             1111                      NaN                NaN        NaN    

          Specialization       How did you hear about X Education  \
77                      Select               Select                 
79      Hospitality Management               Select                 
81                      Select               Select                 
88   Human Resource Management               Select                 
120       Marketing Management               Select                 

    What is your current occupation  \
77                  Unemployed        
79        Working Professional        
81                  Unemployed        
88                  Unemployed        
120       Working Professional        

    What matters most to you in choosing a course Search Magazine  \
77              Better Career Prospects             No      No      
79              Better Career Prospects             No      No      
81              Better Career Prospects             No      No      
88              Better Career Prospects             No      No      
120             Better Career Prospects             No      No      

    Newspaper Article X Education Forums Newspaper Digital Advertisement  \
77          No                No             No              No            
79          No                No             No              No            
81          No                No             No              No            
88          No                No             No              No            
120         No                No             No              No            

    Through Recommendations Receive More Updates About Our Courses  \
77             No                             No                     
79             No                             No                     
81             No                             No                     
88             No                             No                     
120            No                             No                     

                    Tags                    Lead Quality     \
77                                   NaN                NaN   
79   Will revert after reading the email   Low in Relevance   
81                                   NaN                NaN   
88                    Closed by Horizzon  High in Relevance   
120                   Closed by Horizzon   Low in Relevance   

    Update me on Supply Chain Content Get updates on DM Content  \
77                  No                            No              
79                  No                            No              
81                  No                            No              
88                  No                            No          

When Total Time Spent On Website = 0 , Except 3 cases in every case Total Visits is 0.
<p>Total Visits is NA but Total Time Spent On Website != 0 </p>

Action : Impute with mean when Total Time Spent on Website != 0 . NA is just 1.5% of the total distribution.

In [25]:
df[df["Country"].isna()]["City"].value_counts()

City
Select                         1217
Mumbai                          211
Thane & Outskirts                80
Other Cities                     55
Other Cities of Maharashtra      43
Other Metro Cities               15
Tier II Cities                    2
Name: count, dtype: int64

In [26]:
df["Country"].value_counts()

Country
India                   6492
United States             69
United Arab Emirates      53
Singapore                 24
Saudi Arabia              21
United Kingdom            15
Australia                 13
Qatar                     10
Hong Kong                  7
Bahrain                    7
Oman                       6
France                     6
unknown                    5
Kuwait                     4
South Africa               4
Nigeria                    4
Germany                    4
Canada                     4
Sweden                     3
Italy                      2
Belgium                    2
Ghana                      2
Philippines                2
Netherlands                2
China                      2
Asia/Pacific Region        2
Bangladesh                 2
Uganda                     2
Russia                     1
Sri Lanka                  1
Tanzania                   1
Kenya                      1
Malaysia                   1
Liberia                    1
Denmar

In [27]:
df.columns

Index(['Prospect ID', 'Lead Number', 'Lead Origin', 'Lead Source',
       'Do Not Email', 'Do Not Call', 'Converted', 'TotalVisits',
       'Total Time Spent on Website', 'Page Views Per Visit', 'Last Activity',
       'Country', 'Specialization', 'How did you hear about X Education',
       'What is your current occupation',
       'What matters most to you in choosing a course', 'Search', 'Magazine',
       'Newspaper Article', 'X Education Forums', 'Newspaper',
       'Digital Advertisement', 'Through Recommendations',
       'Receive More Updates About Our Courses', 'Tags', 'Lead Quality',
       'Update me on Supply Chain Content', 'Get updates on DM Content',
       'Lead Profile', 'City', 'Asymmetrique Activity Index',
       'Asymmetrique Profile Index', 'Asymmetrique Activity Score',
       'Asymmetrique Profile Score',
       'I agree to pay the amount through cheque',
       'A free copy of Mastering The Interview', 'Last Notable Activity'],
      dtype='object')

In [28]:
df.drop(['Prospect ID', 'Lead Number'] , axis=1 , inplace=True)

<h4>'How did you hear about X Education','Lead Profile',
                'Lead Quality','Asymmetrique Profile Score','Asymmetrique Activity Score',
               'Asymmetrique Activity Index','Asymmetrique Profile Index','Tags','Last Notable Activity' are columns to be dropped because they were found after marketing steps by sales teams</h4>

In [29]:
cols_to_drop = ['How did you hear about X Education','Lead Profile',
                'Lead Quality','Asymmetrique Profile Score','Asymmetrique Activity Score',
               'Asymmetrique Activity Index','Asymmetrique Profile Index','Tags','Last Notable Activity']

df.drop(cols_to_drop, axis=1, inplace = True)

In [30]:
df = df.replace('Select', np.nan)

In [ ]:
categorical_col = df.select_dtypes(exclude =["number"]).columns.values
numerical_col = df.select_dtypes(include =["number"]).columns.values
# print("CATEGORICAL FEATURES : \n {} \n\n".format(categorical_col))
# print("NUMERICAL FEATURES : \n {} ".format(numerical_col))

In [ ]:
# A function that hets information about the categorical information.

def Cat_info(df, categorical_column):
    results = []

    for col in categorical_column:
        values = {
            "columns": col,
            "values": df[col].unique(),
            "unique_values": df[col].nunique(),
            "null_values": df[col].isna().sum(),
            "null_percent": round(df[col].isna().sum() / len(df) * 100, 1)
        }
        results.append(values)

    df_result = pd.DataFrame(results)
    df_result.sort_values("null_values", ascending=False, inplace=True)
    df_result.set_index("columns", inplace=True)
    return df_result


In [33]:
df_cat = Cat_info(df, categorical_col)
df_cat

values                                                                                                                                                                                                           \
columns                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
City                                                                                                                                                                                                                                                                                                                                                    [nan, Mumbai, Thane & Outskirts, Other Metro Cities, Other Cities, Other Cities of Maharashtra, Tier II Cities]   
Specialization                                 [nan, Business Administration, Media and Advertising, Supply Chain Management, IT Projects Management, Finance Management, Travel and Tourism, Human Resource Management, Marketing Management, Banking, Investment And Insurance, International Business, E-COMMERCE, Operations Management, Retail Management, Services Excellence, Hospitality Management, Rural and Agribusiness, Healthcare Management, E-Business]   
What matters most to you in choosing a course                                                                                                                                                                                                                                                                                                                                                          [Better Career Prospects, nan, Flexibility & Convenience, Other]   
What is your current occupation                                                                                                                                                                                                                                                                                                                                                         [Unemployed, Student, nan, Working Professional, Businessman, Other, Housewife]   
Country                                               [nan, India, Russia, Kuwait, Oman, United Arab Emirates, United States, Australia, United Kingdom, Bahrain, Ghana, Singapore, Qatar, Saudi Arabia, Belgium, France, Sri Lanka, China, Canada, Netherlands, Sweden, Nigeria, Hong Kong, Germany, Asia/Pacific Region, Uganda, Kenya, Italy, South Africa, Tanzania, unknown, Malaysia, Liberia, Switzerland, Denmark, Philippines, Bangladesh, Vietnam, Indonesia]   
Last Activity                                                                                            [Page Visited on Website, Email Opened, Unreachable, Converted to Lead, Olark Chat Conversation, Email Bounced, Email Link Clicked, Form Submitted on Website, Unsubscribed, Had a Phone Conversation, View in browser link Clicked, nan, Approached upfront, SMS Sent, Visited Booth in Tradeshow, Resubscribed to emails, Email Received, Email Marked Spam]   
Lead Source                                                                                                                                                                                     [Olark Chat, Organic Search, Direct Traffic, Google, Referral Sites, Welingak Website, Reference, google, Facebook, nan, blog, Pay per Click Ads, bing, Social Media, WeLearn, Click2call, Live Chat, welearnblog_Home, youtubechannel, testone, Press_Release, NC_EDM]   
Do Not Email                                                                                                         

<h4>['Magazine',
 'Receive More Updates About Our Courses',
 'Update me on Supply Chain Content',
 'Get updates on DM Content',
 'I agree to pay the amount through cheque]</h4>

In [34]:
cols_to_drop = df_cat[df_cat['unique_values']==1].index.values.tolist() 
cols_to_drop

['Magazine',
 'Receive More Updates About Our Courses',
 'Update me on Supply Chain Content',
 'Get updates on DM Content',
 'I agree to pay the amount through cheque']

These are having just 1 Unique value and thus dropped

In [35]:
df.drop(cols_to_drop, axis=1, inplace = True)
len(df.columns)

21

In [36]:
categorical_col = df.select_dtypes(exclude =["number"]).columns.values
new_cat = Cat_info(df, categorical_col)
new_cat

values                                                                                                                                                                                                           \
columns                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
City                                                                                                                                                                                                                                                                                                                                                    [nan, Mumbai, Thane & Outskirts, Other Metro Cities, Other Cities, Other Cities of Maharashtra, Tier II Cities]   
Specialization                                 [nan, Business Administration, Media and Advertising, Supply Chain Management, IT Projects Management, Finance Management, Travel and Tourism, Human Resource Management, Marketing Management, Banking, Investment And Insurance, International Business, E-COMMERCE, Operations Management, Retail Management, Services Excellence, Hospitality Management, Rural and Agribusiness, Healthcare Management, E-Business]   
What matters most to you in choosing a course                                                                                                                                                                                                                                                                                                                                                          [Better Career Prospects, nan, Flexibility & Convenience, Other]   
What is your current occupation                                                                                                                                                                                                                                                                                                                                                         [Unemployed, Student, nan, Working Professional, Businessman, Other, Housewife]   
Country                                               [nan, India, Russia, Kuwait, Oman, United Arab Emirates, United States, Australia, United Kingdom, Bahrain, Ghana, Singapore, Qatar, Saudi Arabia, Belgium, France, Sri Lanka, China, Canada, Netherlands, Sweden, Nigeria, Hong Kong, Germany, Asia/Pacific Region, Uganda, Kenya, Italy, South Africa, Tanzania, unknown, Malaysia, Liberia, Switzerland, Denmark, Philippines, Bangladesh, Vietnam, Indonesia]   
Last Activity                                                                                            [Page Visited on Website, Email Opened, Unreachable, Converted to Lead, Olark Chat Conversation, Email Bounced, Email Link Clicked, Form Submitted on Website, Unsubscribed, Had a Phone Conversation, View in browser link Clicked, nan, Approached upfront, SMS Sent, Visited Booth in Tradeshow, Resubscribed to emails, Email Received, Email Marked Spam]   
Lead Source                                                                                                                                                                                     [Olark Chat, Organic Search, Direct Traffic, Google, Referral Sites, Welingak Website, Reference, google, Facebook, nan, blog, Pay per Click Ads, bing, Social Media, WeLearn, Click2call, Live Chat, welearnblog_Home, youtubechannel, testone, Press_Release, NC_EDM]   
Lead Origin                                                                                                          

In [37]:
df.drop("City",axis=1, inplace = True)

<h4>Column :  Specialization</h4/>

In [38]:
df['Specialization'].value_counts(normalize = True)*100

Specialization
Finance Management                   16.655290
Human Resource Management            14.470990
Marketing Management                 14.300341
Operations Management                 8.583618
Business Administration               6.877133
IT Projects Management                6.245734
Supply Chain Management               5.955631
Banking, Investment And Insurance     5.767918
Media and Advertising                 3.464164
Travel and Tourism                    3.464164
International Business                3.037543
Healthcare Management                 2.713311
Hospitality Management                1.945392
E-COMMERCE                            1.911263
Retail Management                     1.706485
Rural and Agribusiness                1.245734
E-Business                            0.972696
Services Excellence                   0.682594
Name: proportion, dtype: float64


It may be possible that the lead has no specialization or may be a student and has no work experience yet , thus he/she has not entered any value. We will create a new category called 'Others' to replace the null values.

In [39]:
df['Specialization'] = df['Specialization'].replace(np.nan, 'Others')

We have replaced Na with others

<h4>Column : What matters most to you in choosing a course</h4>

In [40]:
df['What matters most to you in choosing a course'].value_counts(normalize = True)*100

What matters most to you in choosing a course
Better Career Prospects      99.954065
Flexibility & Convenience     0.030623
Other                         0.015312
Name: proportion, dtype: float64

Too much of skew , Better to delete it 

In [41]:
df.drop('What matters most to you in choosing a course', axis = 1, inplace=True)
len(df.columns)

19

<h4>Column : What is your current occupation</h4>

In [42]:
df['What is your current occupation'].value_counts(normalize=True)*100

What is your current occupation
Unemployed              85.496183
Working Professional    10.778626
Student                  3.206107
Other                    0.244275
Housewife                0.152672
Businessman              0.122137
Name: proportion, dtype: float64

85.5% values are "Unemployed". If we impute the data as "Unemployed" then data will become more skewed. Thus, we will impute the value as "Unknown".

In [43]:
df['What is your current occupation'] = df['What is your current occupation'].replace(np.nan, 'Unknown')
df['What is your current occupation'].value_counts(normalize = True)*100

What is your current occupation
Unemployed              60.606061
Unknown                 29.112554
Working Professional     7.640693
Student                  2.272727
Other                    0.173160
Housewife                0.108225
Businessman              0.086580
Name: proportion, dtype: float64

<h4>Country</h4>

In [44]:
df['Country'].value_counts(normalize=True)

Country
India                   0.957663
United States           0.010178
United Arab Emirates    0.007818
Singapore               0.003540
Saudi Arabia            0.003098
United Kingdom          0.002213
Australia               0.001918
Qatar                   0.001475
Hong Kong               0.001033
Bahrain                 0.001033
Oman                    0.000885
France                  0.000885
unknown                 0.000738
Kuwait                  0.000590
South Africa            0.000590
Nigeria                 0.000590
Germany                 0.000590
Canada                  0.000590
Sweden                  0.000443
Italy                   0.000295
Belgium                 0.000295
Ghana                   0.000295
Philippines             0.000295
Netherlands             0.000295
China                   0.000295
Asia/Pacific Region     0.000295
Bangladesh              0.000295
Uganda                  0.000295
Russia                  0.000148
Sri Lanka               0.000148
Ta

Country data is heavily skewed as 95% of the data is mapped as India. Similar to City, Country data is not required for Model building as X-Education is online platform. We will drop the country columns too.

<h4> Last Activity</h4>

In [45]:
print("Number of null values in Last Activity column is : ", df['Last Activity'].isnull().sum())
print("Percentage of null values in Last Activity column is : ", round(df['Last Activity'].isnull().sum()/df.shape[0]*100,2))

Number of null values in Last Activity column is :  103
Percentage of null values in Last Activity column is :  1.11


In [46]:
df['Last Activity'].value_counts(normalize = True)*100

Last Activity
Email Opened                    37.616285
SMS Sent                        30.042684
Olark Chat Conversation         10.649010
Page Visited on Website          7.004487
Converted to Lead                4.684251
Email Bounced                    3.567911
Email Link Clicked               2.922185
Form Submitted on Website        1.269563
Unreachable                      1.017840
Unsubscribed                     0.667615
Had a Phone Conversation         0.328335
Approached upfront               0.098501
View in browser link Clicked     0.065667
Email Marked Spam                0.021889
Email Received                   0.021889
Resubscribed to emails           0.010945
Visited Booth in Tradeshow       0.010945
Name: proportion, dtype: float64

In [47]:
df['Last Activity'] = df['Last Activity'].replace(np.nan, 'Email Opened')
print("Number of null values in Last Activity column is : ", df['Last Activity'].isnull().sum())

Number of null values in Last Activity column is :  0


As we are unsure what could be the Last activity, we will replace it with the most frequent activity "Email Opened".

<h4>Lead Source</h4>

In [48]:
print("Number of null values in Lead Source column is : ", df['Lead Source'].isnull().sum())
print("Percentage of null values in Lead Source column is : ", round(df['Lead Source'].isnull().sum()/df.shape[0]*100,2))

Number of null values in Lead Source column is :  36
Percentage of null values in Lead Source column is :  0.39


In [49]:
lead = df.copy()

In [50]:
lead['Lead Source'].value_counts(normalize = True)*100

Lead Source
Google               31.160365
Direct Traffic       27.629292
Olark Chat           19.067797
Organic Search       12.538027
Reference             5.801825
Welingak Website      1.542807
Referral Sites        1.358105
Facebook              0.597566
bing                  0.065189
google                0.054324
Click2call            0.043459
Live Chat             0.021730
Social Media          0.021730
Press_Release         0.021730
Pay per Click Ads     0.010865
blog                  0.010865
WeLearn               0.010865
welearnblog_Home      0.010865
youtubechannel        0.010865
testone               0.010865
NC_EDM                0.010865
Name: proportion, dtype: float64

In [51]:
lead['Lead Source'] = lead['Lead Source'].replace(np.nan, 'Google')
lead['Lead Source'] = lead['Lead Source'].replace(['google'], 'Google')
print("Number of null values in Lead Source column is : ", lead['Lead Source'].isnull().sum())

Number of null values in Lead Source column is :  0


As Google seems to be most used Lead Source, we will replace null values with Google. There is a category 'google' which is same as 'Google' We will replace the values

In [52]:
def Num_info(df, numeric_column):
    rows = []
    
    for col in numeric_column:
        null_count = df[col].isna().sum()
        null_percent = (null_count / len(df)) * 100
        rows.append({
            "columns": col,
            "null_values": null_count,
            "null_percent": round(null_percent, 1)
        })
    
    df_result = pd.DataFrame(rows)
    df_result.sort_values("null_values", ascending=False, inplace=True)
    df_result.set_index("columns", inplace=True)
    
    return df_result


In [53]:
df_num = Num_info(lead,numerical_col)
df_num

null_values  null_percent
columns                                               
TotalVisits                      137           1.5    
Page Views Per Visit             137           1.5    
Converted                          0           0.0    
Total Time Spent on Website        0           0.0

<h4>Total Visits</h4>

In [54]:
# Plot
plt.figure(figsize=(6, 3))

# Distribution plot with KDE
plt.subplot(1, 2, 1)
sns.histplot(lead['TotalVisits'].dropna(), kde=True)
plt.title('Distribution of TotalVisits')

# Boxplot
plt.subplot(1, 2, 2)
sns.boxplot(x=lead['TotalVisits'].dropna())
plt.title('Boxplot of TotalVisits')

plt.tight_layout()
plt.show()

As we see there are some outliers in the data, we will impute with median and not mean value.

In [55]:
lead['TotalVisits'].fillna(lead['TotalVisits'].median(), inplace=True)
lead['TotalVisits'].isnull().sum()

np.int64(0)

<h4>Page Views Per Visit</h4>

In [56]:

# Plot
plt.figure(figsize=(6, 3))

# Distribution plot with KDE
plt.subplot(1, 2, 1)
sns.histplot(lead['Page Views Per Visit'].dropna(), kde=True)
plt.title('Distribution of Page Views Per Visit')

# Boxplot
plt.subplot(1, 2, 2)
sns.boxplot(x=lead['Page Views Per Visit'].dropna())
plt.title('Boxplot of Page Views Per Visit')

plt.tight_layout()
plt.show()

In [57]:
lead['Page Views Per Visit'].fillna(lead['Page Views Per Visit'].median(), inplace=True)
lead['Page Views Per Visit'].isnull().sum()

np.int64(0)

As we see there are some outliers in the data, we will impute with median and not mean value. 

In [58]:
df['Lead Origin'].value_counts()

Lead Origin
Landing Page Submission    4886
API                        3580
Lead Add Form               718
Lead Import                  55
Quick Add Form                1
Name: count, dtype: int64

In [59]:
df['Lead Source'].value_counts()

Lead Source
Google               2868
Direct Traffic       2543
Olark Chat           1755
Organic Search       1154
Reference             534
Welingak Website      142
Referral Sites        125
Facebook               55
bing                    6
google                  5
Click2call              4
Live Chat               2
Social Media            2
Press_Release           2
Pay per Click Ads       1
blog                    1
WeLearn                 1
welearnblog_Home        1
youtubechannel          1
testone                 1
NC_EDM                  1
Name: count, dtype: int64